File: src/main/java/net/thauvin/erik/mobibot/commands/tell/Tell.java
/*
 * Tell.java
 *
 * Copyright (c) 2004-2020, Erik C. Thauvin (erik@thauvin.net)
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *   Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *   Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 *   Neither the name of this project nor the names of its contributors may be
 *   used to endorse or promote products derived from this software without
 *   specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package net.thauvin.erik.mobibot.commands.tell;

import net.thauvin.erik.mobibot.Mobibot;
import net.thauvin.erik.mobibot.Utils;
import net.thauvin.erik.mobibot.commands.AbstractCommand;
import net.thauvin.erik.mobibot.commands.links.View;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * The <code>Tell</code> command.
 *
 * @author <a href="https://erik.thauvin.net" target="_blank">Erik C. Thauvin</a>
 * @created 2016-07-02
 * @since 1.0
 */
public class Tell extends AbstractCommand {
    /**
     * Max days property.
     */
    public static final String MAX_DAYS_PROP = "tell-max-days";
    /**
     * Max size proeprty.
     */
    public static final String MAX_SIZE_PROP = "tell-max-size";
    /**
     * The tell command.
     */
    public static final String TELL_CMD = "tell";
    // Arrow
    private static final String ARROW = " --> ";
    // Serialized object file extension
    private static final String SER_EXT = ".ser";
    // All keyword
    private static final String TELL_ALL_KEYWORD = "all";
    //T he delete command.
    private static final String TELL_DEL_KEYWORD = "del";
    // Messages queue
    private final List<TellMessage> messages = new CopyOnWriteArrayList<>();
    // Serialized object file
    private final String serializedObject;
    // Maximum number of days to keep messages
    private int maxDays = 7;
    // Message maximum queue size
    private int maxSize = 50;

    /**
     * Creates a new instance.
     *
     * @param bot The bot.
     */
    public Tell(final Mobibot bot) {
        super(bot);
        initProperties(MAX_DAYS_PROP, MAX_SIZE_PROP);

        // Load the message queue
        serializedObject = bot.getLogsDir() + bot.getName() + SER_EXT;
        messages.addAll(TellMessagesMgr.load(serializedObject, bot.getLogger()));

        if (clean()) {
            save();
        }
    }

    /**
     * Cleans the messages queue.
     *
     * @return <code>true</code> if the queue was cleaned.
     */
    @SuppressWarnings("WeakerAccess")
    final boolean clean() {
        if (getBot().getLogger().isDebugEnabled()) {
            getBot().getLogger().debug("Cleaning the messages.");
        }
        return TellMessagesMgr.clean(messages, maxDays);
    }

    // Delete message.
    private void deleteMessage(final String sender, final String args, final boolean isOp, final boolean isPrivate) {
        final String[] split = args.split(" ");

        if (split.length == 2) {
            final String id = split[1];
            boolean deleted = false;

            if (TELL_ALL_KEYWORD.equalsIgnoreCase(id)) {
                for (final TellMessage message : messages) {
                    if (message.getSender().equalsIgnoreCase(sender) && message.isReceived()) {
                        messages.remove(message);
                        deleted = true;
                    }
                }

                if (deleted) {
                    save();
                    getBot().send(sender, "Delivered messages have been deleted.", isPrivate);
                } else {
                    getBot().send(sender, "No delivered messages were found.", isPrivate);
                }

            } else {
                boolean found = false;

                for (final TellMessage message : messages) {
                    found = message.isMatchId(id);

                    if (found && (message.getSender().equalsIgnoreCase(sender) || getBot().isOp(sender))) {
                        messages.remove(message);

                        save();
                        getBot().send(sender, "Your message was deleted from the queue.", isPrivate);
                        deleted = true;
                        break;
                    }
                }

                if (!deleted) {
                    if (found) {
                        getBot().send(sender, "Only messages that you sent can be deleted.", isPrivate);
                    } else {
                        getBot().send(sender, "The specified message [ID " + id + "] could not be found.", isPrivate);
                    }
                }
            }
        } else {
            helpResponse(args, sender, isOp, isPrivate);
        }
    }

    @NotNull
    @Override
    public String getName() {
        return TELL_CMD;
    }

    @NotNull
    @Override
    public List<String> getHelp() {
        return List.of("To send a message to someone when they join the channel:",
                       Utils.helpIndent("%c " + TELL_CMD + " <nick> <message>"),
                       "To view queued and sent messages:",
                       Utils.helpIndent("%c " + TELL_CMD + ' ' + View.VIEW_CMD),
                       "Messages are kept for " + Utils.bold(maxDays)
                       + Utils.plural(maxDays, " day.", " days."));
    }

    @Override
    public boolean isOp() {
        return false;
    }

    @Override
    public boolean isPublic() {
        return isEnabled();
    }

    @Override
    public boolean isVisible() {
        return isEnabled();
    }

    @Override
    public void commandResponse(@NotNull final String sender,
                                @NotNull final String login,
                                @NotNull final String args,
                                final boolean isOp,
                                final boolean isPrivate) {
        if (isEnabled()) {
            if (StringUtils.isBlank(args)) {
                helpResponse(args, sender, isOp, isPrivate);
            } else if (args.startsWith(View.VIEW_CMD)) {
                if (getBot().isOp(sender) && (View.VIEW_CMD + ' ' + TELL_ALL_KEYWORD).equals(args)) {
                    viewAll(sender, isPrivate);
                } else {
                    viewMessages(sender, isPrivate);
                }
            } else if (args.startsWith(TELL_DEL_KEYWORD + ' ')) {
                deleteMessage(sender, args, isOp, isPrivate);
            } else {
                newMessage(sender, args, isOp, isPrivate);
            }

            if (clean()) {
                save();
            }
        }
    }

    @Override
    public boolean isEnabled() {
        return maxSize > 0 && maxDays > 0;
    }

    @Override
    public void setProperty(@NotNull final String key, @NotNull final String value) {
        super.setProperty(key, value);
        if (MAX_DAYS_PROP.equals(key)) {
            this.maxDays = Utils.getIntProperty(value, maxDays);
        } else if (MAX_SIZE_PROP.equals(key)) {
            this.maxSize = Utils.getIntProperty(value, maxSize);
        }
    }

    // New message.
    private void newMessage(final String sender, final String args, final boolean isOp, final boolean isPrivate) {
        final String[] split = args.split(" ", 2);

        if (split.length == 2 && (StringUtils.isNotBlank(split[1]) && split[1].contains(" "))) {
            if (messages.size() < maxSize) {
                final TellMessage message = new TellMessage(sender, split[0], split[1].trim());

                messages.add(message);

                save();

                getBot().send(sender, "Message [ID " + message.getId() + "] was queued for "
                                      + Utils.bold(message.getRecipient()), isPrivate);
            } else {
                getBot().send(sender, "Sorry, the messages queue is currently full.", isPrivate);
            }
        } else {
            helpResponse(args, sender, isOp, isPrivate);
        }
    }

    /**
     * Saves the messages queue.
     */
    @SuppressWarnings("WeakerAccess")
    final void save() {
        TellMessagesMgr.save(serializedObject, messages, getBot().getLogger());
    }

    /**
     * Checks and sends messages.
     *
     * @param nickname  The user's nickname.
     * @param isMessage The message flag.
     */
    public void send(final String nickname, final boolean isMessage) {
        if (isEnabled() && !nickname.equals(getBot().getNick())) {
            messages.stream().filter(message -> message.isMatch(nickname)).forEach(message -> {
                if (message.getRecipient().equalsIgnoreCase(nickname) && !message.isReceived()) {
                    if (message.getSender().equals(nickname)) {
                        if (!isMessage) {
                            getBot().send(nickname, Utils.bold("You") + " wanted me to remind you: "
                                                    + Utils.reverseColor(message.getMessage()),
                                          true);

                            message.setIsReceived();
                            message.setIsNotified();

                            save();
                        }
                    } else {
                        getBot().send(nickname, message.getSender() + " wanted me to tell you: "
                                                + Utils.reverseColor(message.getMessage()),
                                      true);

                        message.setIsReceived();

                        save();
                    }
                } else if (message.getSender().equalsIgnoreCase(nickname) && message.isReceived()
                           && !message.isNotified()) {
                    getBot().send(nickname,
                                  "Your message "
                                  + Utils.reverseColor("[ID " + message.getId() + ']') + " was sent to "
                                  + Utils.bold(message.getRecipient()) + " on "
                                  + Utils.utcDateTime(message.getReceived()),
                                  true);

                    message.setIsNotified();

                    save();
                }
            });
        }
    }

    /**
     * Checks and sends messages.
     *
     * @param nickname The user's nickname.
     */
    public void send(final String nickname) {
        send(nickname, false);
    }

    /**
     * Returns the messages queue size.
     *
     * @return The size.
     */
    public int size() {
        return messages.size();
    }

    // View all messages.
    private void viewAll(final String sender, final boolean isPrivate) {
        if (!messages.isEmpty()) {
            for (final TellMessage message : messages) {
                getBot().send(sender, Utils.bold(message.getSender()) + ARROW + Utils.bold(message.getRecipient())
                                      + " [ID: " + message.getId() + ", "
                                      + (message.isReceived() ? "DELIVERED" : "QUEUED") + ']',
                              isPrivate);
            }
        } else {
            getBot().send(sender, "There are no messages in the queue.", isPrivate);
        }
    }

    // View messages.
    private void viewMessages(final String sender, final boolean isPrivate) {
        boolean hasMessage = false;

        for (final TellMessage message : messages) {
            if (message.isMatch(sender)) {
                if (!hasMessage) {
                    hasMessage = true;
                    getBot().send(sender, "Here are your messages: ", isPrivate);
                }

                if (message.isReceived()) {
                    getBot().send(sender,
                                  Utils.bold(message.getSender()) + ARROW + Utils.bold(message.getRecipient())
                                  + " [" + Utils.utcDateTime(message.getReceived()) + ", ID: "
                                  + Utils.bold(message.getId()) + ", DELIVERED]",
                                  isPrivate);

                } else {
                    getBot().send(sender,
                                  Utils.bold(message.getSender()) + ARROW + Utils.bold(message.getRecipient())
                                  + " [" + Utils.utcDateTime(message.getQueued()) + ", ID: "
                                  + Utils.bold(message.getId()) + ", QUEUED]",
                                  isPrivate);
                }

                getBot().send(sender, Utils.helpIndent(message.getMessage()), isPrivate);
            }
        }

        if (!hasMessage) {
            getBot().send(sender, "You have no messages in the queue.", isPrivate);
        } else {
            getBot().send(sender, "To delete one or all delivered messages:", isPrivate);
            getBot().send(sender,
                          Utils.helpIndent(Utils.helpFormat(
                                  "%c " + TELL_CMD + ' ' + TELL_DEL_KEYWORD + " <id|" + TELL_ALL_KEYWORD + '>',
                                  getBot().getNick(),
                                  isPrivate)),
                          isPrivate);
            getBot().send(sender,
                          "Messages are kept for " + Utils.bold(maxDays)
                          + Utils.plural(maxDays, " day.", " days."), isPrivate);
        }
    }
}

File: src/main/kotlin/net/thauvin/erik/mobibot/commands/tell/Tell.kt
/*
 * Tell.kt
 *
 * Copyright 2004-2025 Erik C. Thauvin (erik@thauvin.net)
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *   Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *   Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 *   Neither the name of this project nor the names of its contributors may be
 *   used to endorse or promote products derived from this software without
 *   specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package net.thauvin.erik.mobibot.commands.tell

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import net.thauvin.erik.mobibot.Utils.bold
import net.thauvin.erik.mobibot.Utils.bot
import net.thauvin.erik.mobibot.Utils.helpCmdSyntax
import net.thauvin.erik.mobibot.Utils.helpFormat
import net.thauvin.erik.mobibot.Utils.isChannelOp
import net.thauvin.erik.mobibot.Utils.plural
import net.thauvin.erik.mobibot.Utils.reverseColor
import net.thauvin.erik.mobibot.Utils.sendMessage
import net.thauvin.erik.mobibot.Utils.toIntOrDefault
import net.thauvin.erik.mobibot.Utils.toUtcDateTime
import net.thauvin.erik.mobibot.commands.AbstractCommand
import net.thauvin.erik.mobibot.commands.links.View
import org.pircbotx.PircBotX
import org.pircbotx.hooks.events.MessageEvent
import org.pircbotx.hooks.types.GenericMessageEvent
import org.pircbotx.hooks.types.GenericUserEvent

/**
 * Queues a message to be sent to someone when they join or are active on the channel.
 */
@SuppressFBWarnings("BC_BAD_CAST_TO_ABSTRACT_COLLECTION")
class Tell(private val serialObject: String) : AbstractCommand() {
    // Messages queue
    private val messages: MutableList<TellMessage> = mutableListOf()

    // Maximum number of days to keep messages
    private var maxDays = 7

    // Message maximum queue size
    private var maxSize = 50

    /**
     * The tell command.
     */
    override val name = "tell"

    override val isOpOnly: Boolean = false
    override val isPublic: Boolean = isEnabled()
    override val isVisible: Boolean = isEnabled()

    init {
        addHelp(
            "To send a message to someone when they join the channel:",
            helpFormat("%c $name <nick> <message>"),
            "To view queued and sent messages:",
            helpFormat("%c $name ${View.VIEW_CMD}"),
            "Messages are kept for ${maxDays.bold()}" + " day".plural(maxDays.toLong()) + '.'
        )
    }

    /**
     * Cleans the messages queue.
     */
    private fun clean(): Boolean {
        return TellManager.clean(messages, maxDays.toLong())
    }

    override fun commandResponse(channel: String, args: String, event: GenericMessageEvent) {
        if (isEnabled()) {
            when {
                args.isBlank() -> {
                    helpResponse(channel, args, event)
                }

                args.startsWith(View.VIEW_CMD) -> {
                    if (event.isChannelOp(channel) && "${View.VIEW_CMD} $TELL_ALL_KEYWORD" == args) {
                        viewAll(event)
                    } else {
                        viewMessages(event)
                    }
                }

                args.startsWith("$TELL_DEL_KEYWORD ") -> {
                    deleteMessage(channel, args, event)
                }

                else -> {
                    newMessage(channel, args, event)
                }
            }
            if (clean()) {
                save()
            }
        }
    }

    // Delete message.
    private fun deleteMessage(channel: String, args: String, event: GenericMessageEvent) {
        val split = args.split(" ")
        if (split.size == 2) {
            val id = split[1]
            if (TELL_ALL_KEYWORD.equals(id, ignoreCase = true)) {
                if (messages.removeIf { it.sender.equals(event.user.nick, true) && it.isReceived }) {
                    save()
                    event.sendMessage("Delivered messages have been deleted.")
                } else {
                    event.sendMessage("No delivered messages were found.")
                }
            } else {
                if (messages.removeIf {
                        it.id == id && (it.sender.equals(event.user.nick, true)
                                || event.isChannelOp(channel))
                    }) {
                    save()
                    event.sendMessage("The message was deleted from the queue.")
                } else {
                    event.sendMessage("The specified message [ID $id] could not be found.")
                }
            }
        } else {
            helpResponse(channel, args, event)
        }
    }

    override fun isEnabled(): Boolean {
        return maxSize > 0 && maxDays > 0
    }

    override fun setProperty(key: String, value: String) {
        super.setProperty(key, value)
        if (MAX_DAYS_PROP == key) {
            maxDays = value.toIntOrDefault(maxDays)
        } else if (MAX_SIZE_PROP == key) {
            maxSize = value.toIntOrDefault(maxSize)
        }
    }

    // New message.
    private fun newMessage(channel: String, args: String, event: GenericMessageEvent) {
        val split = args.split(" ".toRegex(), 2)
        if (split.size == 2 && split[1].isNotBlank() && split[1].contains(" ")) {
            if (messages.size < maxSize) {
                val message = TellMessage(event.user.nick, split[0], split[1].trim())
                messages.add(message)
                save()
                event.sendMessage("Message [ID ${message.id}] was queued for ${message.recipient.bold()}")
            } else {
                event.sendMessage("Sorry, the messages queue is currently full.")
            }
        } else {
            helpResponse(channel, args, event)
        }
    }

    /**
     * Saves the messages queue.
     */
    private fun save() {
        TellManager.save(serialObject, messages)
    }

    /**
     * Checks and sends messages.
     */
    fun send(event: GenericUserEvent) {
        val nickname = event.user.nick
        if (!isEnabled() || nickname == event.getBot<PircBotX>().nick) return

        messages.filter { it.isMatch(nickname) }.forEach { message ->
            when {
                // Deliver a message to the recipient
                message.recipient.equals(nickname, ignoreCase = true) && !message.isReceived -> {
                    val messageText = if (message.sender == nickname) {
                        // Self-reminder (but not from MessageEvent to avoid loops)
                        if (event is MessageEvent) return@forEach
                        "${"You".bold()} wanted me to remind you: ${message.message.reverseColor()}"
                    } else {
                        // Message from someone else
                        "${message.sender} wanted me to tell you: ${message.message.reverseColor()}"
                    }

                    event.user.send().message(messageText)
                    message.isReceived = true
                    if (message.sender == nickname) message.isNotified = true
                    save()
                }

                // Notify sender that message was delivered
                message.sender.equals(nickname, ignoreCase = true) && message.isReceived
                        && !message.isNotified -> {
                    event.user.send().message(
                        "Your message ${"[ID ${message.id}]".reverseColor()} was sent to " +
                                "${message.recipient.bold()} on ${message.receptionDate}"
                    )
                    message.isNotified = true
                    save()
                }
            }
        }
    }

    /**
     * Returns the messages queue size.
     *
     * @return The size.
     */
    fun size(): Int = messages.size

    // View all messages.
    private fun viewAll(event: GenericMessageEvent) {
        if (messages.isNotEmpty()) {
            for (message in messages) {
                event.sendMessage(
                    "${message.sender.bold()}$ARROW${message.recipient.bold()} [ID: ${message.id}, " +
                            (if (message.isReceived) "DELIVERED]" else "QUEUED]")
                )
            }
        } else {
            event.sendMessage("There are no messages in the queue.")
        }
    }

    // View messages.
    private fun viewMessages(event: GenericMessageEvent) {
        var hasMessage = false
        for (message in messages.filter { it.isMatch(event.user.nick) }) {
            if (!hasMessage) {
                hasMessage = true
                event.sendMessage("Here are your messages: ")
            }
            if (message.isReceived) {
                event.sendMessage(
                    message.sender.bold() + ARROW + message.recipient.bold() +
                            " [${message.receptionDate.toUtcDateTime()}, ID: ${message.id.bold()}, DELIVERED]"
                )
            } else {
                event.sendMessage(
                    message.sender.bold() + ARROW + message.recipient.bold() +
                            " [${message.queued.toUtcDateTime()}, ID: ${message.id.bold()}, QUEUED]"
                )
            }
            event.sendMessage(helpFormat(message.message))
        }
        if (!hasMessage) {
            event.sendMessage("You have no messages in the queue.")
        } else {
            event.sendMessage("To delete one or all delivered messages:")
            event.sendMessage(
                helpFormat(
                    helpCmdSyntax(
                        "%c $name $TELL_DEL_KEYWORD <id|$TELL_ALL_KEYWORD>",
                        event.bot().nick, true
                    )
                )
            )
            event.sendMessage(help.last())
        }
    }

    companion object {
        /**
         * Max number of days property.
         */
        const val MAX_DAYS_PROP = "tell-max-days"

        /**
         * Max size property.
         */
        const val MAX_SIZE_PROP = "tell-max-size"

        // Arrow
        private const val ARROW = " --> "

        // The `all` keyword
        private const val TELL_ALL_KEYWORD = "all"

        // The delete command.
        private const val TELL_DEL_KEYWORD = "del"
    }

    /**
     * Creates a new instance.
     */
    init {
        initProperties(MAX_DAYS_PROP, MAX_SIZE_PROP)

        // Load the message queue
        messages.addAll(TellManager.load(serialObject))
        if (clean()) {
            save()
        }
    }
}
